### PR TITLE
UCP/WIREUP: Filter slow lanes in selection process

### PR DESCRIPTION
**Pull request by shasson5**
## What
Filter slow lanes in selection process.

## Why ?
Avoid selecting TCP lanes if we have IB lanes.
The filtering is done by removing lanes with a large enough progress overhead compared to the best lane.
By default, each lane with more than 0.25 of the minimum overhead is filtered out.

### CODE REVIEW

File: src/ucp/wireup/select.c [hunk 1]
```diff
@@ -1426,29 +1427,94 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+    ucp_worker_iface_t *wiface = ucp_worker_iface(worker, rsc_index);
+    uct_perf_attr_t perf_attr;
+
+    perf_attr.field_mask = UCT_PERF_ATTR_FIELD_OPERATION |
+                           UCT_PERF_ATTR_FIELD_PROGRESS_OVERHEAD;
+
+    /* We use AM_SHORT as a dummy operation because we don't know which
+        protocol will be selected (anyway it has no effect on the value of
+        progress_overhead). */
+    perf_attr.operation = UCT_EP_OP_AM_SHORT;
```
Comment (gleon99) [inline src/ucp/wireup/select.c:1442]: Is it needed at all?

File: src/ucp/wireup/select.c [hunk 2]
```diff
@@ -1426,29 +1427,94 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+        overhead     = ucp_wireup_get_lane_progress_overhead(
+                worker, sinfo_array[sinfo_index].rsc_index);
+        min_overhead = ucs_min(overhead, min_overhead);
+    }
+
+    /* Compare each element to minimum progress overhead and filter only fast lanes */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        overhead = ucp_wireup_get_lane_progress_overhead(
+                worker, sinfo_array[sinfo_index].rsc_index);
+        if ((min_overhead > 0) && ((min_overhead / overhead) < max_ratio)) {
```
Comment (gleon99) [inline src/ucp/wireup/select.c:1479]: Something seems wrong with the logic hear.
`min_overhead` will be `0` if there's at least 1 "fast" iface, so this `if` would never happen?
Also, I would care more about checking the denominator (`overhead`) here to prevent division by 0.

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -1426,29 +1427,94 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+            continue;
+        }
+
+        show_error = (num_lanes == 0);
```
Comment (gleon99) [inline src/ucp/wireup/select.c:1483]: `Align `=`.

File: src/ucp/wireup/select.c [hunk 4]
```diff
@@ -1426,29 +1427,94 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+{
+    ucp_worker_iface_t *wiface = ucp_worker_iface(worker, rsc_index);
+    uct_perf_attr_t perf_attr;
+
+    perf_attr.field_mask = UCT_PERF_ATTR_FIELD_OPERATION |
+                           UCT_PERF_ATTR_FIELD_PROGRESS_OVERHEAD;
+
+    /* We use AM_SHORT as a dummy operation because we don't know which
+        protocol will be selected (anyway it has no effect on the value of
+        progress_overhead). */
```
Comment (brminich) [inline src/ucp/wireup/select.c:1441]: alignment

File: src/ucp/wireup/select.c [hunk 5]
```diff
@@ -1426,29 +1427,94 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+    perf_attr.operation = UCT_EP_OP_AM_SHORT;
+
+    if (uct_iface_estimate_perf(wiface->iface, &perf_attr) != UCS_OK) {
```
Comment (brminich) [inline src/ucp/wireup/select.c:1444]: why can't we rely on just BW? We already filtering lanes  according to the BW, but on a later stage, when RMA BW lanes already setup. Maybe we can move this filtering earlier (here), so slow lanes will not be initialized at all?

File: src/ucp/wireup/select.c [hunk 6]
```diff
@@ -1426,29 +1427,94 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+ucp_wireup_add_fast_lanes(const ucp_wireup_select_params_t *select_params,
+                          const ucp_wireup_select_info_t *sinfo_array,
+                          unsigned num_sinfo, ucp_worker_h worker,
+                          ucp_lane_type_t lane_type,
+                          ucp_wireup_select_context_t *select_ctx)
+{
+    unsigned num_lanes     = 0;
+    double min_overhead    = DBL_MAX;
+    ucp_context_h context  = worker->context;
+    const double max_ratio = 1. / context->config.ext.multi_lane_max_ratio;
```
Comment (brminich) [inline src/ucp/wireup/select.c:1462]: now the same parameter is used as ratio of BW and progress overhead, which is confusing

File: src/uct/api/v2/uct_v2.h [hunk 7]
```diff
@@ -167,6 +170,12 @@ typedef struct {
      */
     double              recv_overhead;
 
+    /**
+     * This is the time spent progressing the request, in seconds.
```
[the remark this answers is not in the document]
Comment (shasson5) [reply src/uct/api/v2/uct_v2.h:174]: why not?

File: src/ucp/wireup/select.c (hunk 5 above)
Comment (shasson5) [reply src/ucp/wireup/select.c:1444]: BW filtering cause some issue, because it also filters IB transport from intra node endpoints (where you have SHM transports).

File: src/ucp/wireup/select.c (hunk 6 above)
Comment (shasson5) [reply src/ucp/wireup/select.c:1462]: maybe we should rename it to some more general purpose parameter?

File: src/ucp/wireup/select.c (hunk 2 above)
Comment (shasson5) [reply src/ucp/wireup/select.c:1479]: correct, it was some leftovers from prev code.

Comment (shasson5) [reply src/ucp/wireup/select.c]: min_overhead is always bigger than overhead, so min_overhead > 0 also implies overhead > 0.
if min_overhead is 0 than we don't want to filter anything, so the condition will never be true.

Comment (gleon99) [reply src/ucp/wireup/select.c:1472]: "min_overhead is always bigger than overhead" - are you sure?
Line 1472:
`min_overhead = ucs_min(overhead, min_overhead);`

File: src/uct/api/v2/uct_v2.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (gleon99) [reply src/uct/api/v2/uct_v2.h:174]: The progress is for `iface`.

File: src/ucp/wireup/select.c (hunk 5 above)
Comment (shasson5) [reply src/ucp/wireup/select.c:1437]: that's an option but wouldn't it be a tailor made solution only for this specific issue?

File: test/gtest/ucp/test_ucp_worker.cc [hunk 8]
```diff
@@ -209,6 +209,12 @@ class test_ucp_worker_discard : public ucp_test {
         }
 
         flush_req = sender().flush_worker_nb(0);
+        if ((flush_req == NULL) &&
```
[the remark this answers is not in the document]
Comment (shasson5) [reply test/gtest/ucp/test_ucp_worker.cc:212]: AFAIR, there was an implicit assumption in this test that there will always be some pending requests on the worker.
But when I filtered TCP it stopped being true

File: src/ucp/wireup/select.c [hunk 9]
```diff
@@ -1426,29 +1427,90 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+    /* Iterate over all elements and calculate minimum progress overhead */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        overhead     = ucp_wireup_get_lane_progress_overhead(
+                worker, sinfo[sinfo_index].rsc_index);
+        min_overhead = ucs_min(overhead, min_overhead);
+    }
+
+    /* Compare each element to minimum progress overhead and filter only fast lanes */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        overhead = ucp_wireup_get_lane_progress_overhead(
```
Comment (gleon99) [inline src/ucp/wireup/select.c:1473]: Assert `overhead` before dividing.

File: src/ucp/wireup/select.c [hunk 10]
```diff
@@ -1457,30 +1519,23 @@ ucp_wireup_add_bw_lanes(const ucp_wireup_select_params_t *select_params,
...
-                                                   bw_info->criteria.lane_type,
-                                                   show_error, select_ctx);
+                                                 remote_dev_bitmap, 0,
+                                                 &sinfo[num_sinfo]);
             if (status != UCS_OK) {
                 break;
             }
 
-            num_lanes++;
+            rsc_index  = sinfo[num_sinfo].rsc_index;
```
Comment (gleon99) [inline src/ucp/wireup/select.c:1534]: Align `=`

File: src/ucp/wireup/select.c [hunk 11]
```diff
@@ -1426,29 +1427,90 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
            (ucs_popcount(md_map) < UCP_MAX_OP_MDS);
 }
 
+static double ucp_wireup_get_lane_progress_overhead(ucp_worker_h worker,
+                                                    ucp_rsc_index_t rsc_index)
+{
+    static const double min_valid_overhead = 10e-9;
```
Comment (gleon99) [inline src/ucp/wireup/select.c:1433]: Align `=`

File: src/ucp/wireup/select.c [hunk 12]
```diff
@@ -1426,29 +1427,90 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+    ucp_worker_iface_t *wiface = ucp_worker_iface(worker, rsc_index);
+    uct_perf_attr_t perf_attr;
+
+    perf_attr.field_mask = UCT_PERF_ATTR_FIELD_PROGRESS_OVERHEAD;
+    if (uct_iface_estimate_perf(wiface->iface, &perf_attr) != UCS_OK) {
+        /* Skip the overhead check if no value was retrieved */
+        return min_valid_overhead;
```
Comment (gleon99) [inline src/ucp/wireup/select.c:1440]: WDYT of simplifying the flow (pseudo-code):
```
overhead = 10e-9
if uct_iface_estimate_perf... = OK:
    overhead = max(overhead, perf_attr...)
return overhead
```

File: test/gtest/ucp/test_ucp_worker.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (gleon99) [reply test/gtest/ucp/test_ucp_worker.cc:214]: As discussed, please add a comment here.

[the remark this answers is not in the document]
Comment (shasson5) [reply test/gtest/ucp/test_ucp_worker.cc:214]: already done in line 212

File: src/ucp/wireup/select.c [hunk 13]
```diff
@@ -1426,29 +1428,96 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+        return 0;
+    }
+
+    /* Iterate over all elements and calculate max BW */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        lane_bw = ucp_wireup_get_lane_bw(&sinfo_array[sinfo_index], worker,
+                                         select_params->address->address_list);
+        max_bw  = ucs_max(lane_bw, max_bw);
+
+        printf("lane_bw %f\n", lane_bw);
```
Comment (rakhmets) [inline src/ucp/wireup/select.c:1472]: Please remove this line.

File: src/ucp/wireup/select.c [hunk 14]
```diff
@@ -1426,29 +1428,96 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+    }
+
+    ucs_assertv(max_bw > 0, "max_bw has illegal value: %f", max_bw);
+
+    /* Compare each element to max BW and filter only fast lanes */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        lane_bw = ucp_wireup_get_lane_bw(&sinfo_array[sinfo_index], worker,
+                                         select_params->address->address_list);
+
+        if ((lane_bw / max_bw) < max_ratio) {
```
Comment (brminich) [inline src/ucp/wireup/select.c:1482]: maybe add ucs_trace here what lane (and on what device) was dropped? Can help debugging issues

File: src/uct/api/v2/uct_v2.h [hunk 15]
```diff
@@ -167,6 +170,12 @@ typedef struct {
      */
     double              recv_overhead;
 
+    /**
+     * This is the time spent progressing the iface, in seconds.
+     * This field is set by the UCT layer.
+     */
+    double              progress_overhead;
+
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:178]: are these changes needed now?

File: src/ucp/wireup/select.c [hunk 16]
```diff
@@ -1426,29 +1426,101 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+    }
+
+    /* Iterate over all elements and calculate max BW */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        lane_bw = ucp_wireup_get_lane_bw(&sinfo_array[sinfo_index], worker,
+                                         select_params->address->address_list);
+        max_bw  = ucs_max(lane_bw, max_bw);
+    }
+
+    ucs_assertv(max_bw > 0, "max_bw has illegal value: %f", max_bw);
```
Comment (rakhmets) [inline src/ucp/wireup/select.c:1472]: Which problem does the assert reveal?
Maybe it is better replace line 1480 by `if (lane_bw < (max_bw * max_ratio)) {`, and get rid of assert, if the only purpose of the assert is to avoid division by zero.

File: src/ucp/wireup/select.c [hunk 17]
```diff
@@ -1426,29 +1427,94 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+        /* Skip the overhead check if no value was retrieved */
+        return 0;
+    }
+
+    return perf_attr.progress_overhead;
+}
+
+static unsigned
+ucp_wireup_add_fast_lanes(const ucp_wireup_select_params_t *select_params,
+                          const ucp_wireup_select_info_t *sinfo_array,
```
[the remark this answers is not in the document]
Comment (gleon99) [reply src/ucp/wireup/select.c:1446]: @shasson5 unresolved

File: src/ucp/wireup/select.c [hunk 18]
```diff
@@ -1426,29 +1426,101 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
...
+    const ucp_wireup_select_info_t *sinfo;
+
+    if (num_sinfo == 0) {
+        return 0;
+    }
+
+    /* Iterate over all elements and calculate max BW */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        lane_bw = ucp_wireup_get_lane_bw(&sinfo_array[sinfo_index], worker,
+                                         select_params->address->address_list);
```
Comment (gleon99) [inline src/ucp/wireup/select.c:1468]: Can put `select_params->address->address_list` in a var (used in 2 loops).

File: src/ucp/wireup/select.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (shasson5) [reply src/ucp/wireup/select.c:1446]: notice I added a local var with the same name (sinfo)

File: test/gtest/ucp/test_ucp_worker.cc [hunk 19]
```diff
@@ -209,6 +209,14 @@ class test_ucp_worker_discard : public ucp_test {
         }
 
         flush_req = sender().flush_worker_nb(0);
+        /* In some cases, there will be nothing to flush
+         * so we need to skip the progress loop */
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_worker.cc:213]: pls word-wrap the text to fill the whole line up to 80 chars

File: src/ucp/wireup/select.c [hunk 20]
```diff
@@ -1442,53 +1445,44 @@ static double ucp_wireup_get_lane_bw(const ucp_wireup_select_info_t *sinfo,
 }
 
 static unsigned
-ucp_wireup_add_fast_lanes(const ucp_wireup_select_params_t *select_params,
-                          const ucp_wireup_select_info_t *sinfo_array,
-                          unsigned num_sinfo, ucp_worker_h worker,
+ucp_wireup_add_fast_lanes(ucp_worker_h worker,
+                          const ucp_wireup_select_params_t *select_params,
+                          ucs_array_t(select_info) *sinfo_array,
```
Comment (yosefe) [inline src/ucp/wireup/select.c:1450]: const

File: src/ucp/wireup/select.c [hunk 21]
```diff
@@ -1505,22 +1499,21 @@ ucp_wireup_add_bw_lanes(const ucp_wireup_select_params_t *select_params,
...
     const ucp_address_entry_t *ae;
     ucs_status_t status;
-    unsigned num_sinfo;
     uint64_t local_dev_bitmap;
     uint64_t remote_dev_bitmap;
     ucp_rsc_index_t dev_index;
     ucp_md_map_t md_map;
     ucp_rsc_index_t rsc_index;
     unsigned addr_index;
+    UCS_ARRAY_DEFINE_ONSTACK(sinfo_array, select_info, UCP_MAX_LANES);
```
Comment (yosefe) [inline src/ucp/wireup/select.c:1514]: this is initialized var; it should be 1st

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
